[MM-16585] Document signing plugins (#360)
mattermost/mattermost-developer-documentation: site/content/extend/plugins/signing-plugins.md

---
title: "Signing Plugins"
date: 2019-09-04T00:00:00-05:00
subsection: Plugins (Beta)
weight: 50
---

> "Plugin Marketplace” is a platform used to distribute common plugins, where MM admins can install plugins without verification. To trust a plugin there needs to be a way to verify the authenticity and integrity of a plugin before being installed. [GPG](https://gnupg.org) tool is used to sign and verify plugins. Every plugin is signed using a private key. Every MM instance should have access to a public key which will be used to verify plugin signatures on installation.

Reviewer: ```suggestion
"Plugin Marketplace” is a platform used to distribute common plugins, where Mattermost admins can install plugins without verification. To trust a plugin, there needs to be a way to verify the authenticity and integrity of a plugin before being installed. [The GNU Privacy Guard (GPG)](https://gnupg.org) tool is used to sign and verify plugins. Every plugin is signed using a private key and every Mattermost instance should have access to a public key which is used to verify plugin signatures on installation.
```

## Key Generation

> Public and private key pair is needed to sign and verify plugins. Private key is used for signing and should be kept in the Vault. Public key is used for verification and can be distributed freely. To generate key pair run following command:

Reviewer: ```suggestion
Public and private key pairs are needed to sign and verify plugins. Private key is used for signing and should be kept in the Vault. Public key is used for verification and can be distributed freely. To generate a key pair, run the following command:
```

```console
> gpg --full-generate-key

Please select what kind of key you want:
   (1) RSA and RSA (default)
   (2) DSA and Elgamal
   (3) DSA (sign only)
   (4) RSA (sign only)
Your selection? 1

RSA keys may be between 1024 and 4096 bits long.
What keysize do you want? (2048) 3072

Requested keysize is 3072 bits

Please specify how long the key should be valid.
         0 = key does not expire
      <n>  = key expires in n days
      <n>w = key expires in n weeks
      <n>m = key expires in n months
      <n>y = key expires in n years
Key is valid for? (0) 1y

Key expires at ...

Is this correct? (y/N) y

GnuPG needs to construct a user ID to identify your key.
Real name: Ali Farooq

Email address: ali_@mattermost.com

Comment:

You selected this USER-ID:
    "Ali Farooq <ali_@mattermost.com>"
Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O

```

> note: key size should be at least 3072 bits.

Reviewer: ```suggestion
Note: key size should be at least 3072 bits.
```

> ### Exporting Private Key

Reviewer: ```suggestion
### Exporting the Private Key
```

> Find the ID of your private key first. The ID is the hexadecimal number.

Reviewer: ```suggestion
Find the ID of your private key first. The ID is a hexadecimal number.
```

```console
> gpg --list-secret-keys
```

> This is your private key keep it secret! Replace XXXXXXXX with your hexadecimal key ID

Reviewer: ```suggestion
This is your private key and you should keep it secret. Replace `XXXXXXXX` with your hexadecimal key ID.
```

```console
> gpg --export-secret-keys XXXXXXXX > ./my-priv-gpg-key.plugin.asc
```

> ### Exporting Public Key

Reviewer: ```suggestion
### Exporting the Public Key
```

> Find the ID of your public key first. The ID is the hexadecimal number.

Reviewer: ```suggestion
Find the ID of your public key first. The ID is a hexadecimal number.
```

```console
> gpg --list-keys
```

> Replace XXXXXXXX with your hexadecimal key ID

Reviewer: ```suggestion
Replace `XXXXXXXX` with your hexadecimal key ID.
```

```console
> gpg --export XXXXXXXX > ./my-pub-gpg-key.plugin.asc
```

> ### Importing Key

Reviewer: ```suggestion
### Importing the Key
```

> If you already have a public and private key pair you can import them to the GPG.

Reviewer: ```suggestion
If you already have a public and private key pair, you can import them to the GPG.
```

```console
> gpg --import ./my-priv-gpg-key.plugin.asc
> gpg --import ./my-pub-gpg-key.plugin.asc
```

## Plugin Signing

> For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file. Replace XXXXXXXX with your hexadecimal key ID and run:

Reviewer: ```suggestion
For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file. Replace `XXXXXXXX` with your hexadecimal key ID and run:
```

```console
> gpg -u XXXXXXXX --verbose --personal-digest-preferences SHA256 --detach-sign com.mattermost.demo-plugin-0.1.0.tar.gz
```
This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig` which is the signature of your plugin.

## Plugin Verification

> Mattermost server should be able to verify plugin before install. Public key has to be added on the server. For this purpose start MM server and run Mattermost command:

Reviewer: ```suggestion
Mattermost server should be able to verify plugins before install. The public key has to be added on the server. For this purpose, start the Mattermost server and run the following command:
```

```console
> mattermost plugin add key my-pub-gpg-key.plugin.asc
```

> note: make sure that your public key file extension is `.plugin.asc`

Reviewer: ```suggestion
Note: make sure that your public key file extension is `.plugin.asc`.
```

> Multiple public keys can be added to the MM server:

Reviewer: ```suggestion
Multiple public keys can be added to the Mattermost server:
```

```console
> mattermost plugin add key my-pk-file1.plugin.asc my-pk-file2.plugin.asc
```

> To list names of all public keys installed on your Mattermost server use:

Reviewer: ```suggestion
To list the names of all public keys installed on your Mattermost server, use:
```

```console
> mattermost plugin keys
```

> To delete public key(s) from your Mattermost server use:

Reviewer: ```suggestion
To delete public key(s) from your Mattermost server, use:
```

---
title: "Signing Plugins"
date: 2019-09-04T00:00:00-05:00
subsection: Plugins (Beta)
weight: 50
---

The Plugin Marketplace allows system administrators to download and install plugins from a central repository. Plugins installed via the Plugin Marketplace must be signed by a public key certificate trusted by the local Mattermost server.

While the server ships with a default certificate used to verify plugins from the default Mattermost plugin marketplace, the server can be configured to trust different certificates and point at a different plugin marketplace. This document outlines the steps for generating a public key certificate and signing plugins for use with a custom plugin marketplace. It assumes access to the [GNU Privacy Guard (GPG)](https://gnupg.org) tool.

## Configuration
Configuring plugin signatures allows finer control over the verification process:
```console
PluginSettings.RequirePluginSignature = true
```

> is used to enforce plugin signature verification. With flag on, only marketplace plugins will be installed and verified. With flag off, customers will be able to install plugins manually without signature verification. Note, that marketplace plugins will still be verified even if flag is off.

Reviewer: ```suggestion
is used to enforce plugin signature verification. With flag on, only marketplace plugins will be installed and verified. With flag off, customers will be able to install plugins manually without signature verification. Note that the marketplace plugins will still be verified even if flag is off.
```

      <n>y = key expires in n years
Key is valid for? (0) 0

Key expires at ...

Is this correct? (y/N) y

GnuPG needs to construct a user ID to identify your key.
Real name: Mattermost Inc

Email address: info@mattermost.com

Comment:

You selected this USER-ID:
    "Mattermost Inc <info@mattermost.com>"
Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O

```
Note: key size should be at least 3072 bits.

### Exporting the Private Key
Find the ID of your private key first. The ID is a hexadecimal number.
```console
> gpg --list-secret-keys
```
This is your private key and should be kept secret. Your hexadecimal key ID will, of course, be different.
```console
> gpg --export-secret-keys F3FACE45E0DE642C8BD6A8E64C7C6562C192CC1F > ./my-priv-key
```

### Exporting the Public Key
Find the ID of your public key first. The ID is a hexadecimal number.
```console
> gpg --list-keys
```
```console
> gpg --export F3FACE45E0DE642C8BD6A8E64C7C6562C192CC1F > ./my-pub-key
```

### Importing the Key
If you already have a public and private key pair, you can import them to the GPG.
```console
> gpg --import ./my-priv-gpg-key
> gpg --import ./my-pub-gpg-key
```

## Plugin Signing

> For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file, run:

Reviewer: ```suggestion
For plugin signing, you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file, run:
```

---
title: "Signing Plugins"
date: 2019-09-04T00:00:00-05:00
subsection: Plugins (Beta)
weight: 50
---

The Plugin Marketplace allows system administrators to download and install plugins from a central repository. Plugins installed via the Plugin Marketplace must be signed by a public key certificate trusted by the local Mattermost server.

> While the server ships with a default certificate used to sign plugins from the default Mattermost plugin marketplace, the server can be configured to trust different certificates and point at a different plugin marketplace. This document outlines the steps for generating a public key certificate and signing plugins for use with a custom plugin marketplace. It assumes access to the [GNU Privacy Guard (GPG)](https://gnupg.org) tool.

Reviewer 2: `used to verify plugins`

```console
> gpg --full-generate-key

Please select what kind of key you want:
   (1) RSA and RSA (default)
   (2) DSA and Elgamal
   (3) DSA (sign only)
   (4) RSA (sign only)
Your selection? 1

RSA keys may be between 1024 and 4096 bits long.
What keysize do you want? (2048) 3072

Requested keysize is 3072 bits

Please specify how long the key should be valid.
         0 = key does not expire
      <n>  = key expires in n days
      <n>w = key expires in n weeks
      <n>m = key expires in n months
      <n>y = key expires in n years
Key is valid for? (0) 0

Key expires at ...

Is this correct? (y/N) y

GnuPG needs to construct a user ID to identify your key.
Real name: Mattermost Inc

Email address: info@mattermost.com

Comment:

You selected this USER-ID:
    "Mattermost Inc <info@mattermost.com>"
Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O

```
Note: key size should be at least 3072 bits.

### Exporting the Private Key
Find the ID of your private key first. The ID is a hexadecimal number.
```console
> gpg --list-secret-keys
```

> This is your private key and you should keep it secret. Your hexadecimal key ID will, of course, be different.

Reviewer 2: nit: `This is your private key and should be kept secret.`

```console
> gpg --export-secret-keys F3FACE45E0DE642C8BD6A8E64C7C6562C192CC1F > ./my-priv-key
```

### Exporting the Public Key
Find the ID of your public key first. The ID is a hexadecimal number.
```console
> gpg --list-keys
```
```console
> gpg --export F3FACE45E0DE642C8BD6A8E64C7C6562C192CC1F > ./my-pub-key
```

### Importing the Key
If you already have a public and private key pair, you can import them to the GPG.
```console
> gpg --import ./my-priv-gpg-key
> gpg --import ./my-pub-gpg-key
```

## Plugin Signing
For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file, run:
```console
> gpg -u F3FACE45E0DE642C8BD6A8E64C7C6562C192CC1F --verbose --personal-digest-preferences SHA256 --detach-sign com.mattermost.demo-plugin-0.1.0.tar.gz
```
This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig`, which is the signature of your plugin.

## Plugin Verification

> Mattermost server will verify plugin signatures downloaded from plugin marketplace. To add custom public keys, start the Mattermost server and run the following command:

Reviewer 2: do you have to run the mattermost server to run the cli?

```console
> mattermost plugin add key my-pub-key
```
Multiple public keys can be added to the Mattermost server:
```console
> mattermost plugin add key my-pk-file1 my-pk-file2
```
To list the names of all public keys installed on your Mattermost server, use:
```console
> mattermost plugin keys
```
To delete public key(s) from your Mattermost server, use:
```console
> mattermost plugin delete key my-pk-file1 my-pk-file2
```

> ### Storing Plugin Signatures

Reviewer 2: I think this part can be left out of the documentation as it pertains to the details of how signing is working. Maybe we can link to the plugin signing design doc and leave it at that? Thoughts?

https://docs.google.com/document/d/1qABE7VEx4k_ZAeh6Ydn4pGbu6BQfZt65x68i2s65MOQ/edit#

---
title: "Signing Plugins"
date: 2019-09-04T00:00:00-05:00
subsection: Plugins (Beta)
weight: 50
---

> "Plugin Marketplace” is a platform used to distribute common plugins where Mattermost admins can install plugins without verification. To trust a plugin, there needs to be a way to verify the authenticity and integrity of a plugin before being installed. [The GNU Privacy Guard (GPG)](https://gnupg.org) tool is used to sign and verify plugins. Every plugin is signed using a private key and every Mattermost instance should have access to a public key which is used to verify plugin signatures on installation.

Reviewer 3: ```suggestion
The Plugin Marketplace allows system administrators to download and install plugins from a central repository. Plugins installed via the Plugin Marketplace must be signed by a public key certificate trusted by the local Mattermost server. 

While the server ships with a default certificate used to sign plugins from the default Mattermost plugin marketplace, the server can be configured to trust different certificates and point at a different plugin marketplace. This document outlines the steps for generating a public key certificate and signing plugins for use with a custom plugin marketplace. It assumes access to the [GNU Privacy Guard (GPG)](https://gnupg.org) tool.
```

## Key Generation

> Public and private key pairs are needed to sign and verify plugins. Private key is used for signing and should be kept in the Vault. Public key is used for verification and can be distributed freely. To generate a key pair, run the following command:

Reviewer 3: Let's avoid assuming a vault, and just talk about keeping the private key secure.

>  ```console

Reviewer 3: This block seems to be indented by one white space character?

 > gpg --full-generate-key

 Please select what kind of key you want:
    (1) RSA and RSA (default)
    (2) DSA and Elgamal
    (3) DSA (sign only)
    (4) RSA (sign only)
 Your selection? 1

 RSA keys may be between 1024 and 4096 bits long.
 What keysize do you want? (2048) 3072

 Requested keysize is 3072 bits

 Please specify how long the key should be valid.
          0 = key does not expire
       <n>  = key expires in n days
       <n>w = key expires in n weeks
       <n>m = key expires in n months
       <n>y = key expires in n years
 Key is valid for? (0) 0

 Key expires at ...

 Is this correct? (y/N) y

 GnuPG needs to construct a user ID to identify your key.
 Real name: Mattermost Inc

 Email address: info@mattermost.com

 Comment:

 You selected this USER-ID:
     "Mattermost Inc <info@mattermost.com>"
 Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O

 ```
 Note: key size should be at least 3072 bits.

### Exporting the Private Key
Find the ID of your private key first. The ID is a hexadecimal number.
```console
> gpg --list-secret-keys
```

> This is your private key and you should keep it secret. Replace `XXXXXXXX` with your hexadecimal key ID.

Reviewer 3: It seems we have to repeat the instructions on `XXXXXXXX` multiple times in this document. I'd suggest actually just running the steps, showing the output, and using the value in that output in all the examples below, with a single line around here stating that your keys will, of course, be different.

```console
> gpg --export-secret-keys XXXXXXXX > ./my-priv-key
```

> 

Reviewer 3: nit: sometimes there's two line breaks between headers, sometimes one. I'd suggest just a single line break.

### Exporting the Public Key
Find the ID of your public key first. The ID is a hexadecimal number.
```console
> gpg --list-keys
```
Replace `XXXXXXXX` with your hexadecimal key ID.
```console
> gpg --export XXXXXXXX > ./my-pub-key
```

### Importing the Key
If you already have a public and private key pair, you can import them to the GPG.
```console
> gpg --import ./my-priv-gpg-key
> gpg --import ./my-pub-gpg-key
```

## Plugin Signing
For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file. Replace `XXXXXXXX` with your hexadecimal key ID and run:
```console
> gpg -u XXXXXXXX --verbose --personal-digest-preferences SHA256 --detach-sign com.mattermost.demo-plugin-0.1.0.tar.gz
```
This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig`, which is the signature of your plugin.

## Plugin Verification

> Mattermost server will verify plugin signatures downloaded from Mattermost Marketplace. Mattermost public key is shipped with the server. If needed one can download and verify plugins from custom marketplace as well. To do this, custom public key has to be added on the server. For this purpose, start the Mattermost server and run the following command:

Reviewer 3: Let's use the term `plugin marketplace` to avoid talking about the Mattermost-hosted one, specifically. Also, I proposed a change above that gives the context up front, so we can trim this to just focus on the "how".

```console
> mattermost plugin add key my-pub-key
```

Multiple public keys can be added to the Mattermost server:
```console
> mattermost plugin add key my-pk-file1 my-pk-file2
```
To list the names of all public keys installed on your Mattermost server, use:
```console
> mattermost plugin keys
```
To delete public key(s) from your Mattermost server, use:
```console
> mattermost plugin delete key my-pk-file1 my-pk-file2
```

### Storing Plugin Signatures

> When a plugin is installed, Mattermost server will store plugin signatures in its internal filestore (not to be confused with the public keys). This is to allow the server to verify signatures during the plugin-synchronization process that happens during server startup. To avoid name collisions and overwriting existing files, Mattermost will store plugin signatures alongside plugin bundles using the naming convention mentioned below. The benefit of storing signature files independently is to avoid double-tar, where the tar file contains the bundle tar along with its signature. This would have required Mattermost to migrate existing plugin bundles into double-tar format. Which would have resulted in increased complexity and made upgrade/downgrade scenarios a lot harder to deal with.

Reviewer 3: I don't think we need to talk about `double-tar` and a feature we never shipped.

```console
# Filestore plugin structure
/plugins/<plugin_id>.tar.gz            # plugin bundle tar
/plugins/<plugin_id>.<counter-1>.sig   # multiple plugin signatures
/plugins/<plugin_id>.<counter-2>.sig     

# Examples
/plugins/com.mattermost.demo.tar.gz    # plugin bundle tar
/plugins/com.mattermost.demo.1.sig     # multiple plugin signatures
/plugins/com.mattermost.demo.2.sig
/plugins/com.mattermost.demo.3.sig
/plugins/com.github.matterpoll.tar.gz
/plugins/com.github.matterpoll.1.sig
/plugins/com.github.matterpoll.2.sig
/plugins/com.github.matterpoll.3.sig

```

> ### Mattermost Plugin Assets Naming Convention

Reviewer 3: This seems unrelated to the scope of this document. Perhaps this might be better located in the marketplace documentation itself for how we expect the generator to work.

In order to keep plugin signatures close to their bundle tar, Mattermost will generate plugin signatures and upload them directly to GitHub releases using the following naming convention. This is purely a convention which makes it easier for a human to parse signature and public key pairs at a quick glance.
```console
# Mattermost plugin naming convention
<plugin_id>-<plugin_version>.tar.gz
<plugin_id>-<plugin_version>-<counter>-<public_key_fingerprint_hash>.sig

# Mattermost demo plugin example
com.mattermost.demo-plugin-0.2.0.tar.gz
com.mattermost.demo-plugin-0.2.0-1-1E11.sig
com.mattermost.demo-plugin-0.2.0-2-6E1D.sig
com.mattermost.demo-plugin-0.2.0-3-3MN6.sig
```

> ### Configuration

Reviewer 3: I'd suggest pulling this up earlier in the document to draw emphasis to the high-level details before diving in too deep.

> ---

Reviewer 4: I'm wondering if this documentation makes sense here. I don't see value in this being public (2/5) and it being alongside developer focused documentation (4/5). Maybe move it into https://github.com/mattermost/internal-docs or https://docs.mattermost.com/guides/core.html?

Reviewer 4: This is still an outstanding concern for me

Reviewer 2: @hanzei were you thinking we should wait before we finalize our certification process? This is intentionally put here to guide developers on how to sign their plugins. Some parts I agree can be moved/removed but a guide is necessary for them to know how to sign their own plugin, no?

Reviewer 4: This documentation is needed if you
1. Run your own marketplace (We don't have amazing documentation for this and don't know anyone that wan't to do this right now)
2. Want to also sign your custom plugin for your own marketplace

I highly doubt that someone wants to do this in the near future. And if this becomes the case, we should have a more in debt documentation anyway.

Author: To take step back, we should answer these questions:
1. What is Plugin marketplace?
2. Why did we introduce plugin signatures? And how do they work?
3. What is the process of adding a plugin in the Mattermost Plugin Marketplace? And how can a developer add their plugin to the marketplace?
4. How can customers deploy their custom marketplace?

I think this concrete document should answer question 2. I'm 0/5 whether this document should be here or somewhere else.

Reviewer 4: 1. is answered by https://docs.mattermost.com/administration/plugins.html#plugin-marketplace
3. Is part of an ongoing discussion. I'm unsure where it should be documented.
4. should be answered at https://docs.mattermost.com/administration/plugins.html

With all of this in mind I would opt to move this to docs.mattermost.com. The audience is mostly admins and docs.mattermost.com are the admin docs.

Reviewer 4: :+1: for adding this to https://docs.mattermost.com/administration/plugins.html#security

Reviewer 5: Great. I'll add this to the existing docs once it's been merged.

Do we have any documentation answering "why and how" wrt the signatures?

Reviewer 4: Well, this document kind of answers these two questions.

Reviewer 5: Great - that's what I'd assumed but always best to make sure. :)

Reviewer 5: Is this PR ready to be merged? No rush; but would like to make sure it's moved into the Admin docs as soon as possible. :)

Author: I think it is ready. @aaronrothschild thoughts?

Reviewer 4: Why do you want to merge this PR if we want to have the documentation in the admin docs anyway? Can't we just resubmit this PR to the admin docs?

Reviewer 5: Good question, @hanzei. My thinking is that it's neater to merge this PR and then take the content and move it into the admin guide once it's had all comments/issues/conversations resolved. I could be approaching it the wrong way, so happy to discuss further. :)

title: "Signing Plugins"
date: 2019-09-04T00:00:00-05:00
subsection: Plugins (Beta)
weight: 50
---

> "Plugin Marketplace” is a platform used to distribute common plugins, where Mattermost admins can install plugins without verification. To trust a plugin, there needs to be a way to verify the authenticity and integrity of a plugin before being installed. [The GNU Privacy Guard (GPG)](https://gnupg.org) tool is used to sign and verify plugins. Every plugin is signed using a private keyand every Mattermost instance should have access to a public key which is used to verify plugin signatures on installation.

Reviewer: ```suggestion
"Plugin Marketplace” is a platform used to distribute common plugins where Mattermost admins can install plugins without verification. To trust a plugin, there needs to be a way to verify the authenticity and integrity of a plugin before being installed. [The GNU Privacy Guard (GPG)](https://gnupg.org) tool is used to sign and verify plugins. Every plugin is signed using a private keyand every Mattermost instance should have access to a public key which is used to verify plugin signatures on installation.
```

   (2) DSA and Elgamal
   (3) DSA (sign only)
   (4) RSA (sign only)
   (7) DSA (set your own capabilities)
   (8) RSA (set your own capabilities)
   (9) ECC and ECC
  (10) ECC (sign only)
  (11) ECC (set your own capabilities)
  (13) Existing key

Your selection? 9

Please select which elliptic curve you want:
   (1) Curve 25519
   (3) NIST P-256
   (4) NIST P-384
   (5) NIST P-521
   (6) Brainpool P-256
   (7) Brainpool P-384
   (8) Brainpool P-512
   (9) secp256k1

Your selection? 1

Please specify how long the key should be valid.
         0 = key does not expire
      <n>  = key expires in n days
      <n>w = key expires in n weeks
      <n>m = key expires in n months
      <n>y = key expires in n years

Key is valid for? (0) 0

Key expires at ...

Is this correct? (y/N) y

GnuPG needs to construct a user ID to identify your key.

Real name: Mattermost Inc

Email address: info@mattermost.com

Comment:
You selected this USER-ID:
    "Mattermost Inc <info@mattermost.com>"

Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O
```

### Exporting the Private Key
Find the ID of your private key first. The ID is a hexadecimal number.
```console
> gpg --list-secret-keys
```

> This is your private key and you should keep it secret. Replace `XXXXXXXX` with your hexadecimal key ID

Reviewer: ```suggestion
This is your private key and you should keep it secret. Replace `XXXXXXXX` with your hexadecimal key ID.
```

```console
> gpg --export-secret-keys XXXXXXXX > ./my-priv-key.plugin.gpg
```

### Exporting the Public Key
Find the ID of your public key first. The ID is a hexadecimal number.
```console
> gpg --list-keys
```

> Replace `XXXXXXXX` with your hexadecimal key ID

Reviewer: ```suggestion
Replace `XXXXXXXX` with your hexadecimal key ID.
```

```console
> gpg --export XXXXXXXX > ./my-pub-key.plugin.gpg
```

### Importing the Key
If you already have a public and private key pair, you can import them to the GPG.
```console
> gpg --import ./my-priv-gpg-key.plugin.gpg
> gpg --import ./my-pub-gpg-key.plugin.gpg
```

## Plugin Signing
For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file. Replace `XXXXXXXX` with your hexadecimal key ID and run:
```console
> gpg -u XXXXXXXX --verbose --personal-digest-preferences SHA256 --detach-sign com.mattermost.demo-plugin-0.1.0.tar.gz
```

> This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig` which is the signature of your plugin.

Reviewer: ```suggestion
This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig`, which is the signature of your plugin.
```

## Plugin Verification

> Mattermost server should be able to verify plugins before install. the public key has to be added on the server. For this purpose, start the Mattermost server and run the following command:

Reviewer 4: It's not clear to me that this is only needed, if you want to sign and verify plugins your own, which you only need, if you are running a marketplace yourself.

```console

> > mattermost plugin add key my-pub-key.plugin.gpg

Reviewer 4: 1/5 that we should not encourage people to use `plugin.gpg` as an extensions, as this is an internal convention.

Author: True. We are not. We will automatically add the `plugin.gpg` extension if it's missing.

```

> Note: if your public key file does not have an extension `.plugin.gpg` Mattermost will append it automatically.

Reviewer: ```suggestion
Note: if your public key file does not have an extension `.plugin.gpg`, Mattermost will append it automatically.
```

Multiple public keys can be added to the Mattermost server:
```console
> mattermost plugin add key my-pk-file1.plugin.gpg my-pk-file2.plugin.gpg
```
To list the names of all public keys installed on your Mattermost server, use:
```console
> mattermost plugin keys
```
To delete public key(s) from your Mattermost server, use:
```console
> mattermost plugin delete key my-pk-file1.plugin.gpg my-pk-file2.plugin.gpg
```

> ### Storing plugin signatures

Reviewer: ```suggestion
### Storing Plugin Signatures
```

> When a plugin is installed, Mattermost server will store plugin signatures in its internal filestore(not to be confused with the public keys). This is to allow the server to verify signatures during the plugin-synchronization process that happens during server startup. To avoid name collisions and overwriting existing files, Mattermost will store plugin signatures along side plugin bundles using the naming convention mentioned below. The benefit of storing signature files independently is to avoid double-tar, where the tar file contains the bundle tar along with its signature. This would have required Mattermost to migrate existing plugin bundles into double-tar format. Which would have resulted in increased complexity and made upgrade/downgrade scenarios a lot harder to deal with.

Reviewer: ```suggestion
When a plugin is installed, Mattermost server will store plugin signatures in its internal filestore (not to be confused with the public keys). This is to allow the server to verify signatures during the plugin-synchronization process that happens during server startup. To avoid name collisions and overwriting existing files, Mattermost will store plugin signatures alongside plugin bundles using the naming convention mentioned below. The benefit of storing signature files independently is to avoid double-tar, where the tar file contains the bundle tar along with its signature. This would have required Mattermost to migrate existing plugin bundles into double-tar format, which would have resulted in increased complexity and made upgrade/downgrade scenarios a lot harder to deal with.
```

```console
# Filestore plugin structure
/plugins/<plugin_id>.tar.gz            # plugin bundle tar
/plugins/<plugin_id>.<counter-1>.sig   # multiple plugin signatures
/plugins/<plugin_id>.<counter-2>.sig     

# Examples
/plugins/com.mattermost.demo.tar.gz    # plugin bundle tar
/plugins/com.mattermost.demo.1.sig     # multiple plugin signatures
/plugins/com.mattermost.demo.2.sig
/plugins/com.mattermost.demo.3.sig
/plugins/com.github.matterpoll.tar.gz
/plugins/com.github.matterpoll.1.sig
/plugins/com.github.matterpoll.2.sig
/plugins/com.github.matterpoll.3.sig

```

> ## Mattermost plugin assets naming convention

Reviewer: ```suggestion
## Mattermost Plugin Assets Naming Convention
```

> In order to keep plugin signatures close to their bundle tar, Mattermost will generate plugin signatures and upload them directly to github releases using the following naming convention. This is purely a convention which makes it easier for a human to parse signature and public key pairs at a quick glance.

Reviewer: ```suggestion
In order to keep plugin signatures close to their bundle tar, Mattermost will generate plugin signatures and upload them directly to GitHub releases using the following naming convention. This is purely a convention which makes it easier for a human to parse signature and public key pairs at a quick glance.
```
